File: python/Project Euler/42_triangleWords.py
#!/usr/bin/python

from TriangleWords_42 import *

# 14 is the max characters
# so max triangle word can be 14 * 26 = 364 < sigma(26)

def get_triangleNumbers ():
	i = 1
	tri_list = []
	running_sum = 0
	while ( i <= 26 ):
		running_sum += i
		tri_list.append(running_sum)

		i += 1

	return tri_list

def sum_of_chars (word):
	sum_chars = 0
	for i in word:
		sum_chars += (ord(i) - ord('A') + 1)
	
	return sum_chars

def addifTriangleWord (word, triangleNums):
	if sum_of_chars(word) in triangleNums:
		return 1

	return 0

triangleNums = get_triangleNumbers()
print reduce ((lambda x, y: x + addifTriangleWord(y, triangleNums)), inWords, 0)
# 162

File: haskell/project_euler/42_triangleWords.hs
import IO
import TriangleWords_42
import Char

{-
14 is the max characters
so max triangle word can be 14 * 26 = 364 < sigma(26)
-}

triangleNumbers :: [Int]
triangleNumbers = map (\x -> (x*(x+1)) `div` 2) [1..26]

sum_of_chars :: String -> Int
sum_of_chars inword = foldl (\x y -> x + (ord(y) - ord('A')) + 1) 0 inword

num_triangle_words = 
	length (filter (\x -> (elem (sum_of_chars x) triangleNumbers)) inWords)

main = do
	print num_triangle_words 
	{- 162 -}
